### ISSUE TITLE
Support symmetric encryption in raw targets

### ISSUE THREAD

**Issue by faern** [digint/btrbk#157]
It would be very nice to add the option to encrypt raw backups with symmetric encryption using OpenSSL. Symmetric encryption are thought to be more future proof and more quantum resistant. In a case where you trust your local computer but not where you store your backups it might makes sense to store the encryption secret locally in order to obtain higher security for the data actually uploaded somewhere less safe. This is of course up each and every one and their threat model.

Symmetric encryption is what QubesOS uses. In their current release they do something like this:
`openssl enc -e -$CIPHER -pass $PASS` as the command. Where `$CIPHER` can be configured and be anything listed in `openssl enc -help`, `aes-256-cbc` would be a reasonable default. `$PASS` is fixed to "pass:$PASSPHRASE" in Qubes, but would be nice if it would accept anything allowed to send to the `-pass` flag, more on this under "PASS PHRASE ARGUMENTS" in `man openssl`.

However, for the upcoming release, QubesOS started using `scrypt` instead. I have not investigated this further yet, but I suppose `scrypt` is to be trusted since they switched to that. There it's possible to do `scrypt enc - -` to simply make it read stdin and write on stdout. Here the passphrase must be given to the pty of the process, and that involves some more work. How QubesOS does it can be found here:

https://github.com/QubesOS/qubes-core-admin/blob/master/qubes/backup.py#L254

Personally I would probably write a small wrapper myself working kind of like cron. That way I can make my wrapper ask for the symmetric key when I start it and hold that in memory, then my wrapper can spawn btrbk in regular intervals and provide the key. That way my key never touches the disk, but I would have to load my wrapper script manually on each reboot, which is totally fine for me.

I would love add this support myself and submit a PR. But I don't know any Perl, so that would take some time. If someone better at Perl added this I would appreciate it very much. Otherwise I might get around to try to do it myself at some point.

**Comment by digint** [comment]
AFAIK key derivation functions (KDF) are less secure than public key crypto. `openssl enc` uses a modified PBKDF2 algorithm, while `scrypt` pretends to be more secure, but it's still KDF.

Instead, I'd rather go for a **true symmetric approach**, which can be done using openssl:

### Proposal

1. Add new options for `btrbk.conf`, something like:
```
raw_target_compress  xz
raw_target_encrypt   aes-256-cbc
raw_target_keyfile   btrbk_aes256.key

volume /mnt/btr_pool
  subvolume home
    target raw ssh://cloud.example.com/backup
```

2. Generate secret key on the source host, and store it beneath the subvolumes to be encrypted:

```
KEYFILE=/mnt/btr_pool/btrbk_aes256.key
dd if=/dev/urandom bs=1 count=32 | od -x -A n | tr -d "[:space:]" > $KEYFILE
```

3. btrbk command would look something like this:
```
IV=`dd if=/dev/urandom bs=1 count=16 | od -x -A n | tr -d "[:space:]"`

btrfs send <snapshot> \
| xz -c \
| openssl enc -aes-256-cbc -iv $IV -K `cat <raw_target_keyfile>` \
| ssh btrbk@cloud.example.com 'dd of=/backup/home.20170613'

echo -e "cipher=aes256\niv=$IV" \
| ssh btrbk@cloud.example.com 'dd of=/backup/home.20170613.info'
```

4. Add some goodies like `btrbk keygen` which qould automate step 2, as it's always a good thing to rotate the keys from time to time.


#### Further thoughts

For some use cases it might be suitable to define the key directly in btrbk.conf, e.g.:
```
raw_target_key  63490144c2dbf8783582d275f36f1b8a12c0785114b1c2a28a120c4ba57a6f4d
```

**Comment by faern** [comment]
As far as I know KDF is only weak for short passphrases. If using anything with sufficient entropy they should be perfectly fine. The docs should probably point that out to people though.

One property I wanted to have was to never let the key touch the disk.

Your solution has the benefit of not requiring manual start and key entry and that a badly picked passphrase will not jeopardize security. I'm not sure exactly what I prefer now.

**Comment by digint** [comment]
> One property I wanted to have was to never let the key touch the disk.

I'd say putting a key beneath the actual file system is ok, as an attacker who can read the key will also be able to read the filesystem (this also holds if you use filesystem-encryption, as long as the key is within the encrypted filesystem).

Some valid arguments to use KDF are if you want to quickly create an emergency backup, or maybe if you want to share your backups with others without needing to securely transmit key data. For long-term backups, you'll probably write down the passphrase anyway.

Actually also adding KDF to btrbk should not be much more complex than my proposal above, the only thing is that I always wanted to avoid having user-interaction within the program.

Also note that I don't have much free time for btrbk coding at the moment, and implementing all this will probably lead to quite some rewrite of the (pretty hacky) raw-target code.

**Comment by faern** [comment]
For long term backups I'm quite likely to use the same passphrase I use for my full disk encryption, or a variant thereof. Thus one will not forget it, because it's used on every boot.

**Comment by faern** [comment]
Note that you don't need to require user interactions within btrbk itself. You can just make it pick up the passphrase from a filedescriptor or something and the parent process can provide it like that. But yes, that requires the parent process to be something smarter than just cron, as I stated in the original issue.

I understand if you don't have much free time. It's the same with me and most of my hobby projects. Just wanted to add this issue here in case you or anyone have time and the knowledge to implement it :)

**Comment by DL6AKU** [comment]
Just wanted to add that aes-cbc / aes-xts or other "block mode" aes is probably not the right way to do this.

> If you're encrypting files, and not disk sectors, don't use XTS. Use an AEAD scheme, like GCM or OCB or Salsa20+Poly1305. If you're working with files, you have format flexibility. Use it. Randomize your encryption. Authenticate it with a proper MAC tag.

https://sockpuppet.org/blog/2014/04/30/you-dont-want-xts/

> Encrypting data: Use AES in CTR (Counter) mode, and append an HMAC.
AES is about as standard as you can get, and has done a good job of resisting cryptologic attacks over the past decade. Using CTR mode avoids the weakness of ECB mode, the complex (and bug-prone) process of padding and unpadding of partial blocks (or ciphertext stealing), and vastly reduces the risk of side channel attacks thanks to the fact that the data being input to AES is not sensitive. However, because CTR mode is malleable, you should always add an HMAC to confirm that the encrypted data has not been tampered with.
In some situations it may be preferable for performance reasons to use a cipher mode such as GCM which combines encryption and authentication; but this benefit is small (HMACs are fast) and increases the risk of side channel attacks (because attacker-supplied input is processed).

http://www.daemonology.net/blog/2009-06-11-cryptographic-right-answers.html

**Comment by digint** [comment]
well yes, ciphers...

The idea is that the user can choose whatever he wants, but of course we'll need to provide good defaults. I used aes-cbc in the example above mainly because I used this in an old project before, and I know it works.

Here's some things I recall / tested:

- CBC
  - uses padding, and was broken lately with known plaintext attacks (not sure if this also holds for btrfs send streams, but at least some of the data is known).
  - needs an IV
  - needs separate HMAC if you want authenticated encryption
- CTR:
  - needs a nonce (tests using `openssl enc` showed that it does not generate the nonce by itself, so we must provide a noce to the send-stream ourselves, which is not practicable)
  - needs separate HMAC if you want authenticated encryption
- GCM:
  - needs an IV
  - provides authenticated encryption
  - maximum size of 64GB ([link RFC5116](https://tools.ietf.org/html/rfc5116#page-14))
  - not supported with `openssl enc` ("AEAD ciphers not supported by the enc utility")
- XTS
  - complicated and not recommended by crypto experts
  - not supported with `openssl enc` ("Ciphers in XTS mode are not supported by the enc utility")
- CCM
  - similar to GCM, but worse (broken?)
- others
  - not sure, but as far as I know not recommended

#### Authenticated Encryption

Not necessarily needed for our puropse, but nice to have. The problem is that we have a stream, and it might get problematic to HMAC after/before encrypting (probably best scheme woud be "encrypt-then-mac")

#### Command line tools

There's very little (I know of).

- `openssl enc`: provides symmetric encryption with many ciphers/modes, as well as KDF
- gnupg: also provides symmetric encryption (`gpg --symmetric`), but only with KDF (probably PBKDF2?)
- [pcp](https://github.com/TLINDEN/pcp): uses libsodium (yum!), provides asymmetric encryption using CURVE25519 (yum!), as well as symmetric KDF with scrypt (which probably wins over PBKDF2 for being more future-proof)

**Comment by DL6AKU** [comment]
```openssl enc``` does support ```-chacha20```, so if not authentication is wanted, this should be the best option.

    $ echo lol|openssl enc -chacha20 -e -k foobar |openssl enc -chacha20 -d -k foobar|hexdump -C
    00000000  6c 6f 6c 0a                                       |lol.|
    00000004

**Comment by DL6AKU** [comment]
Something like this script that I quickly wrote. Not sure about the right scrypt params though.

    #!/usr/bin/env python3
    
    import hashlib
    import sys
    import os
    
    if len(sys.argv) <= 1:
    	print("Usage: {} <password>".format(sys.argv[0]))
    	sys.exit(1)
    
    salt = os.urandom(32)
    key = hashlib.scrypt(' '.join(sys.argv[1:]).encode("utf-8"), salt=salt, n=2**16, r=8, p=1, maxmem=2**28, dklen=32)
    
    salt = ["{:02X}".format(x) for x in salt]
    key = ["{:02X}".format(x) for x in key]
    
    print("Salt:", ' '.join(salt))
    print("Key: ", ' '.join(key))

**Comment by digint** [comment]
Cool,  -chacha20 is supported in openssl-v1.1.[?]. This makes `openssl enc` (for now) a winner for me. Will take a closer look at "[link pcp](https://github.com/TLINDEN/pcp)" also, this looks pretty sleek and promising.

I still run `dev-libs/openssl-1.0.2k::gentoo` here, `openssl enc -h` does not list `-chacha20`.

### Proposal

make `-chacha20` default, and offer `-aes-256-cbc` as a fallback or openssl-v1.0.x, which (i think) lots of distros still have in "stable". (experts can still specify their own cipher/mode by config options).

Separate KDF is a second thing. It's getting problematic for remote sources, as you want to do all the crypto there. one-time (keygen) passwd entry via ssh seems also feasable, so maybe it's easily doable.

---

pcp test (after generating key):
```
# echo https://github.com/TLINDEN/pcp > plain.txt 
# cat plain.txt | bin/pcp1 -e --sign --z85-encode -r axel > cipher.txt.pcp
Enter passphrase to decrypt your secret key: 
Encrypted 180 bytes for:
  0x516072EF687D2F3F - axel <axel@tty0.ch>
Signed encrypted file successfully
# cat cipher.txt.pcp
----- BEGIN PCP ENCRYPTED FILE -----
7YTKo0EY+KW/}:Qdbz28D2*Jhf)D60U*4H0^$z^zxIt*>[mUt]]YeK4Iz]u5@]6gibOm-eF
}P6]5p-J1s9!XZkv#b3@%9acIu9-ca/w9<qlvLrQ[LA!BZ>oWzlAbLXo3v@K{Z(!Ap7W?H$
/^s6-jJ-D0C*)i.c1PB9![A{9E12Pn-*G1Ul7E{Lu$OSO0dTnwxKbK%shfAR:=mrXRe%kK8
V85qAC6J)Rq3n<gU#8/v6lvp8Zl2w6]]i.0z9=Z8/k=tcfB5YPT$AoTo/xdhX{}@*{}!>AM
I?7pK.>L4OuBt3p4w5!C58{W56I+6rF-lBPdob&8uFm.#wQ&hwknaG}0Egk$O8Qp}[MyPS&
^%J)GjZ}6<!99SWpf{SY3V&S]zH$:k&JkCs9KcvTg^3l/
----- END PCP ENCRYPTED FILE -----
# cat cipher.txt.pcp | bin/pcp1 -d --z85-decode --check-signature
Enter passphrase to decrypt your secret key: 
https://github.com/TLINDEN/pcp
Decrypted and Verified 31 bytes successfully
```

**Comment by DL6AKU** [comment]
$ openssl version 
    OpenSSL 1.1.0f  25 May 2017

**Comment by digint** [comment]
**There's always time to implement crypto!**

See new hot [branch openssl_enc](/digint/btrbk/tree/openssl_enc), especially commit 2025f59ee3d4209908f60ade0971383ee3f2fa18.

It's still experimental, but seems to work fine!


#### Step 1: create key:
```
# KEYFILE=/tmp/btrbk_unittest/mnt_source/btrbk.key
# dd if=/dev/urandom bs=1 count=32 | od -x -A n | tr -d "[:space:]" > $KEYFILE
```

#### Step 2: btrbk config

btrbk.conf.openssl_enc:
```
transaction_log /tmp/btrbk_unittest/tlog

volume /tmp/btrbk_unittest/mnt_source
  incremental no
  raw_target_encrypt openssl_enc
  openssl_ciphername aes-256-cbc
  openssl_iv_size    16
  openssl_keyfile    /tmp/btrbk_unittest/mnt_source/btrbk.key

  subvolume svol
    target raw /tmp/btrbk_unittest/mnt_target/
```

#### Step 3: create backups

```
# ./btrbk -l debug  -c btrbk.conf.openssl_enc run
[...]
Creating full backup...
Generating iv for openssl encryption (cipher=aes-256-cbc)
16+0 records in
16+0 records out
16 bytes copied, 6.936e-05 s, 231 kB/s
[send-to-raw] source: /tmp/btrbk_unittest/mnt_source/svol.20170616
[send-to-raw] target: /tmp/btrbk_unittest/mnt_target/svol.20170616.btrfs.encrypted
Creating raw info file (complete): /tmp/btrbk_unittest/mnt_target/svol.20170616.btrfs.encrypted.info
### echo -e -n "#btrbk-v0.25.1-dev\n# Do not edit this file\nTYPE=raw\nFILE=svol.20170616.btrfs.encrypted\nRECEIVED_UUID=c4b7507e-64d3-4a40-bb90-d23a7d8450a4\nINCOMPLETE=1\nencrypt=openssl_enc\ncipher=aes-256-cbc\niv=29c259727d811d9f93fd818a58593b61\nsource_url=/tmp/btrbk_unittest/mnt_source/svol.20170616\ntarget_url=/tmp/btrbk_unittest/mnt_target/svol.20170616.btrfs.encrypted\n" > /tmp/btrbk_unittest/mnt_target/svol.20170616.btrfs.encrypted.info
Command execution successful
### btrfs send -v /tmp/btrbk_unittest/mnt_source/svol.20170616 | openssl enc -e -aes-256-cbc -K $(cat /tmp/btrbk_unittest/mnt_source/btrbk.key) -iv 29c259727d811d9f93fd818a58593b61 | dd status=none bs=128K of=/tmp/btrbk_unittest/mnt_target/svol.20170616.btrfs.encrypted
At subvol /tmp/btrbk_unittest/mnt_source/svol.20170616
BTRFS_IOC_SEND returned 0
joining genl thread
Command execution successful
[...]
--------------------------------------------------------------------------------
Backup Summary (btrbk command line client, version 0.25.1-dev)

    Date:   Fri Jun 16 17:22:17 2017
    Config: btrbk.conf.openssl_enc

Legend:
    ===  up-to-date subvolume (source snapshot)
    +++  created subvolume (source snapshot)
    ---  deleted subvolume
    ***  received subvolume (non-incremental)
    >>>  received subvolume (incremental)
--------------------------------------------------------------------------------
/tmp/btrbk_unittest/mnt_source/svol
+++ /tmp/btrbk_unittest/mnt_source/svol.20170616_3
*** /tmp/btrbk_unittest/mnt_target/svol.20170616.btrfs.encrypted
```

#### Step 4: verify (decrypt)

```
# KEYFILE=/tmp/btrbk_unittest/mnt_source/btrbk.key
# BACKUP=/tmp/btrbk_unittest/mnt_target/svol.20170616.btrfs.encrypted
# KEY=`cat $KEYFILE`
# IV=`sed -n 's/^iv=//p' ${BACKUP}.info`

# openssl enc -d -aes-256-cbc -K $KEY -iv $IV -in $BACKUP | hexdump -C | head -n1
00000000  62 74 72 66 73 2d 73 74  72 65 61 6d 00 01 00 00  |btrfs-stream....|
```

**Comment by DL6AKU** [comment]
Wow, that was fast!

Sorry, I can't test it this weekend. 

> /tmp/btrbk_unittest/mnt_source/btrbk.key

Just make sure you don't actually put your keyfile in ```/tmp``` ;). While this method is simple, it is virtually impossible to access your backups if you lose the keyfile.

It would be nice if in the future you could derive a key and use that with a salt. I had not thought about just saving the salt in the name of the backup file, that would be a nice option. While a 256 bit salt would be nice, this would result in a long string:

    e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855

It does not look much better in base64 (with ```base64.b64encode(0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.to_bytes(32, byteorder='big'), altchars=b'_-')```

    47DEQpj8HBSa_-TImW_5JCeuQeRkm5NMpJWZG3hSuFU=

You could use a 64 bit salt and I guess that would be OK if the passphrase has enough entropy.

    e3b0c44298fc1c14
    47DEQpj8HBQ=

Or you could use the ```username@host``` as a salt. But then you have to tell the users that they have to remember username@host. Again, passphrase entropy should be high.

**Comment by faern** [comment]
I like the derived keys approach. The point of a backup is to back up the stuff I can't keep in my head, such as long random keys :P Which is why I want the key to my backups in my head only, to avoid a circular dependency.

**Comment by digint** [comment]
Implemented generic KDF support, using external backends: af7cd1a2c294a1191540ff28f1e01c49d0b842d1, on [branch openssl_enc](/digint/btrbk/tree/openssl_enc).

It's all still pretty experimental, I used Python (with my pretty poor skills) to implement a pbkdf2 backend (preferred hashlib.scrypt does not work for me, I think this needs openssl-1.1).

The backends take the derived-key-length as argument, and output the key as well as other (generic) key/value pairs:

```
# contrib/crypt/kdf_pbkdf2.py 32
Passphrase: 
Retype passphrase: 
KEY=df9279b42951277bfc0f4de98b7ba5c9b4308840ed72a07f74e93d2b9fd3823c
algoritm=pbkdf2_hmac
hash_name=sha256
salt=01b6e44b4128c789288191acb48a0c8b
iterations=300000
```

The KEY is then used by btrbk for encrypting (via `openssl enc`), and the extra values are appended with `kdf_` prefix to the raw backup sidecar file (`<backup_file>.info`).

#### Example:

btrbk.conf.openssl_enc:
```
transaction_log /tmp/btrbk_unittest/tlog

volume /tmp/btrbk_unittest/mnt_source
  incremental no
  raw_target_encrypt openssl_enc
  openssl_ciphername aes-256-cbc
  openssl_iv_size    16  # NOTE: set to "no" if no IV is needed by the selected cipher
  #  openssl_keyfile    /tmp/btrbk_unittest/mnt_source/btrbk.key

  # only absolute accepted for now:
  kdf_backend    /home/axel/devel/btrbk/contrib/crypt/kdf_pbkdf2.py

  # accepts "once" (backend is run only once) or "each" (backend is run for each backup)
  kdf_keygen     once

  subvolume svol
    target raw /tmp/btrbk_unittest/mnt_target/
```

run btrbk:
```
# ./btrbk -v -c btrbk.conf.openssl_enc run
btrbk command line client, version 0.25.1-dev  (Mon Jun 26 15:54:53 2017)
[...]
Creating subvolume backup (send-receive) for: /tmp/btrbk_unittest/mnt_source/svol.20170626
Creating full backup...
Generating iv for openssl encryption (cipher=aes-256-cbc)
Passphrase: 
Retype passphrase: 
[send-to-raw] source: /tmp/btrbk_unittest/mnt_source/svol.20170626
[send-to-raw] target: /tmp/btrbk_unittest/mnt_target/svol.20170626.btrfs.encrypted
[...]
--------------------------------------------------------------------------------
/tmp/btrbk_unittest/mnt_source/svol
+++ /tmp/btrbk_unittest/mnt_source/svol.20170626
*** /tmp/btrbk_unittest/mnt_target/svol.20170626.btrfs.encrypted


# cat /tmp/btrbk_unittest/mnt_target/svol.20170626.btrfs.encrypted.info 
#btrbk-v0.25.1-dev
# Do not edit this file
TYPE=raw
FILE=svol.20170626.btrfs.encrypted
RECEIVED_UUID=dc2d9515-b41e-0644-8f7e-c92268982437
encrypt=openssl_enc
cipher=aes-256-cbc
iv=6657a28c9cc675a1573c9070127b49d4
kdf_algoritm=pbkdf2_hmac
kdf_hash_name=sha256
kdf_iterations=300000
kdf_salt=6c450ce04a56d75e9bb446d44259f616
```

**Comment by digint** [comment]
Cleaned/squashed/amended "openssl_enc" branch (now obsolete), and merged master into new [raw_symmetric_encryption](/digint/btrbk/tree/openssl_enc) branch.

Have been running it in vacation (push-to-cloud :) as well as daily backups, and it proves quite stable.

Will probably merge it into 0.26.0, which will require to run `contrib/migration/raw_suffix2sidecar` [1] on existent raw targets.

  [1] https://github.com/digint/btrbk/blob/raw_symmetric_encryption/contrib/migration/raw_suffix2sidecar

**Comment by digint** [comment]
Merged "raw_symmetric encryption" branch (now obsolete) to master, with some fixes. See commits de7628ac7caec037d5c726d5050777347db7d5b7 and 422d52c063a16527b0e186e1212214fa10117390.

Leaving this issue open for further discussion, as it is all still very experimental.

**Comment by digint** [comment]
included in btrbk-0.26.0 (tagged as "very experimental" in documentation)

**Comment by calestyo** [comment]
I've just stumbled over this when looking at the symmetric encryption features... and didn't understand from the documentation how e.g. `kdf_pbkdf2.py` is actually called, and where it gets its data from (the "passphrase" - which I guess is in `openssl_keyfile`?) and the KDF parameters?

I even already wanted to ask for Argon2 support, since PBKDF2 does't protect so well anymore.

But having thought a little while over it, wonder, whether using a KDF makes sense in the `btrbk` case at all:

* What we really need is a high entropy key for the symmetric encryption.<br>
  And we can get that easily from `/dev/random`.
* For automated backups, that key anyway needs to be stored on disk ... or at least read from somewhere else (e.g. from a secure remote ssh server).
* In any case though, there will be some place, where it's stored. Especially, from the `btrbk` point of view, this place is **not** the brain of some human.<br>
   How the key itself is backuped, whether it's written on some paper, or generated by a KDF from a memorable password, is beyond the scope of `btrbk`.

So in my opinion, a KDF isn't needed at all here, it just wastes computing power during the derivation, because the passphrase and parameters for the KDF would still need to be stored somewhere, where the host can get them - but then, we can also directly store&use a high entropy passphrase (i.e. the key).

The "only" thing one needs a KDF for, is to convert a low entropy passphrase (i.e. something that humans can easily remember) into something semi high entropy, i.e. to make "test_1_am_re@lly_bad_in_Pa$$w0rds" into something not easily brute forceable.

But when we anyway store the weak passphrase somewhere, we can as well just store some good key.

The only other reason for a KDF is to make an arbitrary passphrase (even if high entropy) fit the needed key(size):<br>
E.g. the `diceware` created passphrase:
```
SuperbowlStepSanctionDallyingChuckRevivingEaselSidingSandwormRecantCyclicComfortVaporizerOverhungVillageAppeaseCyclistPruneBrokerLiningDeviationCaravanAutomaticDenimPrudeZoologistSwingingMothballFanciedShufflingAnimateDenialShopliftCrimpEmptierCarvingExonerateBrokerDoorpostEngine
```
should be extremely high entropy... but wouldn't fit e.g. a 256 Bit key of AES256.

But for that, purpose, simple cryptographically secure has should also do, AFAIU (though I'm not a crypto expert).


Cheers,
Chris.

**Comment by faern** [comment]
> For automated backups, that key anyway needs to be stored on disk ... or at least read from somewhere else (e.g. from a secure remote ssh server).

I don't agree. See this paragraph from my original issue post:

Personally I would probably write a small wrapper myself working kind of like cron. That way I can make my wrapper ask for the symmetric key when I start it and hold that in memory, then my wrapper can spawn btrbk in regular intervals and provide the key. That way my key never touches the disk, but I would have to load my wrapper script manually on each reboot, which is totally fine for me.

**Comment by calestyo** [comment]
Okay but this modus operand seems then to be rather targeted towards really interactive use (i.e. there is a human who manually enters the passphrase - not just some automated way to fetch it from somewhere remote[0]) - at least once (per boot) when it needs to ask for the passphrase, right?

I guess that makes some sense for (single) systems...where the passphrase is entered by human.

But then it would IMO be advisable to migrate the script to Argon2id.


btw:
- Just reading it "normally" into memory still has some chance for it to end up on disk (if swap is used).
- Also since it was mentioned further up, OpenPGP does not use PBKFD2... nor (at least in current version) Argon2 or anything else that can be considered modern. It uses its own "S2K", which is IMO rather weak. Shouldn't be a problem though with public/private key encryption, only when using low entropy passphrases with symmetric key encryption.

[0] Sure, one could fetch it (the passphrase) in some automated fashion from a remote server and then only store it in memory (e.g. kernel keyring)... but in that case again, KDF makes no sense, as the remote side could just readily provide a high entropy passphrase/key.

**Comment by faern** [comment]
> seems then to be rather targeted towards really interactive use

Equally interactive as dropbear SSH for unlocking disks on boot. Can be a headless server that you want to protect as much as possible, and you are fine to tend to it once at boot to enter credentials.

**Comment by digint** [comment]
@calestyo symmetric encryption is implemented, see btrbk.conf(5) `openssl_enc` option. This will run `openssl enc`, and used to work fine (the configuration of it might need some more love though).

Regarding Argon2: The reason why argon2 is not implemented was that it was not widely available at the time of writing. It should be straight forward to implement, similar to:
https://github.com/digint/btrbk/blob/master/contrib/crypt/kdf_pbkdf2.py

https://github.com/digint/btrbk/blob/master/btrbk#L1790

**Comment by calestyo** [comment]
> @calestyo symmetric encryption is implemented, see btrbk.conf(5) openssl_enc option. This will run openssl enc, and used to work fine (the configuration of it might need some more love though).

Sure, that’s clear. The point I'm trying to make is, that for most cases (i.e. all where they passphrase/key is not manually entered by a human, but e.g. read from disk or via SSH), it doesn't really make much sense to use a KDF, as one could just directly provide a high entropy (and then the KDF is just useless - it doesn’t harm either, apart from wasting energy perhaps).

**Comment by calestyo** [comment]
Oh and regarding Argon2, if a CLI tool shall be used for that, there’s the "official" `argon2` (but at least [in Debian this is built in an annoying way - without threads](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1014110).)
